File: src/chipschallenge/blockreactions/BombTo.java
package chipschallenge.blockreactions;

import chipschallenge.Block;
import chipschallenge.SoundPlayer.sounds;

/**
 * Move to bomb
 */
public class BombTo extends BlockReaction {

    private BombTo() {
    }
    private static BombTo mInstance = null;

    public static synchronized BombTo getInstance() {
        if (mInstance == null) {
            mInstance = new BombTo();
        }
        return mInstance;
    }

    // If collision, both bomb and other block is removed
    public void react(Block moving, Block standing) {
        if (moving.isChip()) {
            die("Ooops! Don't touch the bombs!");
        } else {
            moving.destroy();
            standing.destroy();
            sound().playSound(sounds.BOMB);
        }
    }

    // All can move here
    public boolean canMove(Block moving, Block standing) {
        return true;
    }
}

File: ChipsChallenge.Shared/Blockreactions/BombTo.cs
namespace ChipsChallenge.Shared.Blockreactions
{
    public class BombTo : NoSlipReaction
    {
        private static BombTo instance;

        private BombTo()
        {
        }

        public static BombTo Instance
        {
            get
            {
                lock (typeof(BombTo))
                {
                    return instance ?? (instance = new BombTo());
                }
            }
        }


        public override void React(Block moving, Block standing)
        {
            if (moving.Chip)
            {
                Die("Ooops! Don't touch the bombs!");
            }
            else
            {
                // If collision, both bomb and other block is removed
                moving.Destroy();
                standing.Destroy();
                Sound().Play(Shared.Sound.Bomb);
            }
        }

        public override bool canMove(Block moving, Block standing)
        {
            return true;
        }
    }
}